simplicity-sys/src/ffi.rs

```diff
@@ -0,0 +1,480 @@
+            -30 => Err(SimplicityErr::Cmr),
+            -32 => Err(SimplicityErr::Amr),
+            -34 => Err(SimplicityErr::ExecBudget),
+            -36 => Err(SimplicityErr::ExecMemory),
+            -38 => Err(SimplicityErr::ExecJet),
+            -40 => Err(SimplicityErr::ExecAssert),
+            -42 => Err(SimplicityErr::AntiDoS),
+            x => panic!("unexpected error code {}", x),
+        }
+    }
+}
+
+pub mod bitstream {
+    use super::bitstring::CBitstring;
+    use libc::{c_int, c_uchar, size_t};
+
+    /// Stream of bits.
+    #[repr(C)]
+    #[derive(Copy, Clone)]
+    pub struct CBitstream {
+        pub arr: *const c_uchar,
+        pub len: size_t,
+        pub offset: c_uchar,
+    }
+
+    impl From<&[u8]> for CBitstream {
+        fn from(sl: &[u8]) -> CBitstream {
+            CBitstream {
+                arr: sl.as_ptr(),
+                len: sl.len(),
+                offset: 0,
+            }
+        }
+    }
+
+    extern "C" {
+        pub fn closeBitstream(stream: *mut CBitstream) -> bool;
+        pub fn readNBits(n: c_int, stream: *mut CBitstream) -> i32;
+        pub fn decodeUptoMaxInt(stream: *mut CBitstream) -> i32;
+        pub fn readBitstring(result: *mut CBitstring, n: size_t, stream: *mut CBitstream) -> i32;
+    }
+}
+
+pub mod bitstring {
+    use libc::{c_uchar, size_t};
+
+    /// String of bits.
+    #[repr(C)]
+    #[derive(Copy, Clone)]
+    pub struct CBitstring {
+        arr: *const c_uchar,
+        len: size_t,
+        offset: size_t,
+    }
+
+    impl Default for CBitstring {
+        fn default() -> Self {
+            CBitstring {
+                arr: std::ptr::null(),
+                len: 0,
+                offset: 0,
+            }
+        }
+    }
+}
+
+pub mod dag {
+    use super::bitstring::CBitstring;
+    use super::sha256::CSha256Midstate;
+    use super::ty::CType;
+    use super::SimplicityErr;
+    use libc::{c_void, size_t};
+
+    /// Kind of Simplicity node.
+    #[repr(C)]
+    #[derive(Copy, Clone, Eq, PartialEq, Debug)]
+    pub enum CTag {
+        COMP,
+        CASE,
+        ASSERTL,
+        ASSERTR,
+        PAIR,
+        DISCONNECT,
+        INJL,
+        INJR,
+        TAKE,
+        DROP,
+        IDEN,
+        UNIT,
+        HIDDEN,
+        WITNESS,
+        JET,
+        WORD,
+    }
+
+    /// Used to count the different kinds of combinators in a Simplicity DAG.
+    #[repr(C)]
+    #[derive(Copy, Clone, Eq, PartialEq, Debug, Default)]
+    pub struct CCombinatorCounters {
+        pub comp_cnt: size_t,
+        pub case_cnt: size_t,
+        pub pair_cnt: size_t,
+        pub disconnect_cnt: size_t,
+        pub injl_cnt: size_t,
+        pub injr_cnt: size_t,
+        pub take_cnt: size_t,
+        pub drop_cnt: size_t,
+    }
+
+    /// A source and target type (in the C code this is actually an anonymous struct
+    /// where it is used).
+    #[repr(C)]
+    #[derive(Copy, Clone, Eq, PartialEq, Debug)]
+    pub struct CArrow {
```

Comment:
This is unused in FFI. But okay to keep here.

Reply:
It's actually not used at all. I will remove it.